lib/hsa/staging_buffer.cpp

```diff
@@ -0,0 +1,242 @@
+void StagingBuffer::CopyHostToDevice(void* dst, const void* src, size_t sizeBytes, hsa_signal_t *waitFor)
+{
+    std::lock_guard<std::mutex> l (_copy_lock);
+
+    const char *srcp = static_cast<const char*> (src);
+    char *dstp = static_cast<char*> (dst);
+
+    for (int i=0; i<_numBuffers; i++) {
+        hsa_signal_store_relaxed(_completion_signal[i], 0);
+    }
+
+    if (sizeBytes >= UINT64_MAX/2) {
+        THROW_ERROR (hipErrorInvalidValue);
+    }
+    int bufferIndex = 0;
+    for (int64_t bytesRemaining=sizeBytes; bytesRemaining>0 ;  bytesRemaining -= _bufferSize) {
+
+        size_t theseBytes = (bytesRemaining > _bufferSize) ? _bufferSize : bytesRemaining;
+
+        tprintf (DB_COPY2, "H2D: waiting... on completion signal handle=%lu\n", _completion_signal[bufferIndex].handle);
+        hsa_signal_wait_acquire(_completion_signal[bufferIndex], HSA_SIGNAL_CONDITION_LT, 1, UINT64_MAX, HSA_WAIT_STATE_ACTIVE);
+
+        tprintf (DB_COPY2, "H2D: bytesRemaining=%zu: copy %zu bytes %p to stagingBuf[%d]:%p\n", bytesRemaining, theseBytes, srcp, bufferIndex, _pinnedStagingBuffer[bufferIndex]);
+        // TODO - use uncached memcpy, someday.
+        memcpy(_pinnedStagingBuffer[bufferIndex], srcp, theseBytes);
+
+
+        hsa_signal_store_relaxed(_completion_signal[bufferIndex], 1);
+
+        hsa_status_t hsa_status = hsa_amd_memory_async_copy(dstp, _hsa_agent, _pinnedStagingBuffer[bufferIndex], _cpu_staging_agent, theseBytes, waitFor ? 1:0, waitFor, _completion_signal[bufferIndex]);
+        tprintf (DB_COPY2, "H2D: bytesRemaining=%zu: async_copy %zu bytes %p to %p status=%x\n", bytesRemaining, theseBytes, _pinnedStagingBuffer[bufferIndex], dstp, hsa_status);
+
+        if (hsa_status != HSA_STATUS_SUCCESS) {
+            THROW_ERROR ((hipErrorRuntimeMemory));
+        }
+
+        srcp += theseBytes;
+        dstp += theseBytes;
+        if (++bufferIndex >= _numBuffers) {
+            bufferIndex = 0;
+        }
+
+        // Assume subsequent commands are dependent on previous and don't need dependency after first copy submitted, HIP_ONESHOT_COPY_DEP=1 
+        waitFor = NULL; 
+    }
+
+
+    for (int i=0; i<_numBuffers; i++) {
+        hsa_signal_wait_acquire(_completion_signal[i], HSA_SIGNAL_CONDITION_LT, 1, UINT64_MAX, HSA_WAIT_STATE_ACTIVE);
+    }
+}
+
+//---
+//Copies sizeBytes from src to dst, using either a copy to a staging buffer or a staged pin-in-place strategy
+//IN: dst - dest pointer - must be accessible from agent this buffer is associated with (via _hsa_agent).
+//IN: src - src pointer for copy.  Must be accessible from host CPU.
+//IN: waitFor - hsaSignal to wait for - the copy will begin only when the specified dependency is resolved.  May be NULL indicating no dependency.
+void StagingBuffer::CopyDeviceToHost(void* dst, const void* src, size_t sizeBytes, hsa_signal_t *waitFor)
```

Comment:
Sizet is already uint64- how would change to int64  help?

On Sep 1, 2016, at 4:55 AM, Wen-Heng (Jack) Chung <notifications@github.com<mailto:notifications@github.com>> wrote:


In lib/hsa/staging_buffer.cpp<https://github.com/RadeonOpenCompute/hcc/pull/115#discussion_r77145548>:

> +        // Assume subsequent commands are dependent on previous and don't need dependency after first copy submitted, HIP_ONESHOT_COPY_DEP=1
> +        waitFor = NULL;
> +    }
> +
> +
> +    for (int i=0; i<_numBuffers; i++) {
> +        hsa_signal_wait_acquire(_completion_signal[i], HSA_SIGNAL_CONDITION_LT, 1, UINT64_MAX, HSA_WAIT_STATE_ACTIVE);
> +    }
> +}
> +
> +//---
> +//Copies sizeBytes from src to dst, using either a copy to a staging buffer or a staged pin-in-place strategy
> +//IN: dst - dest pointer - must be accessible from agent this buffer is associated with (via _hsa_agent).
> +//IN: src - src pointer for copy.  Must be accessible from host CPU.
> +//IN: waitFor - hsaSignal to wait for - the copy will begin only when the specified dependency is resolved.  May be NULL indicating no dependency.
> +void StagingBuffer::CopyDeviceToHost(void* dst, const void* src, size_t sizeBytes, hsa_signal_t *waitFor)


@sunway513<https://github.com/sunway513> , @bensander<https://github.com/bensander> should we change the interface so sizeBytes is int64_t , so we can move larger data buffers between host and device?

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/RadeonOpenCompute/hcc/pull/115/files/b7fce5bfd878c26940a3b80bdc53443bbfb80fc3#r77145548>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ACYSAl_KY07vQ5evylTyOCZtTr59GkcOks5qlqEfgaJpZM4Jsak9>.